### PR TITLE
[NeMo-UX] async checkpointing support

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -392,10 +396,23 @@ def _save_checkpoint(self, trainer: 'pytorch_lightning.Trainer', filepath: str)
                 super()._save_checkpoint(trainer, filepath)
             self.remove_checkpoint_unfinished_marker(filepath, barrier_before=True)
         else:
+            # Async save passed the finalization function to checkpoint_io,
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/callbacks/model_checkpoint.py:399]:
```suggestion
            # Async save passes the finalization function to checkpoint_io,
```

File: nemo/lightning/io/pl.py [hunk 2]
```diff
@@ -8,12 +8,26 @@
 from lightning_fabric.plugins.io.checkpoint_io import CheckpointIO
 from lightning_fabric.utilities.cloud_io import get_filesystem
 from lightning_fabric.utilities.types import _PATH
+from megatron.core.dist_checkpointing.serialization import (
+    get_default_load_sharded_strategy,
+    get_default_save_sharded_strategy,
+)
+
+# from nemo.utils.callbacks.torch_dist_async import TorchDistAsyncSaveShardedStrategy
```
Comment (mikolajblaz) [inline nemo/lightning/io/pl.py:16]:
```suggestion
```

File: nemo/lightning/pytorch/optim/megatron.py [hunk 3]
```diff
@@ -90,10 +90,9 @@ def sharded_state_dict(
                 model_sharded_state_dict,
                 optimizer_state_dict=None,
                 is_loading=False,
-                # dist_ckpt_parallel_save=False, ## TODO: fix!
+                dist_ckpt_parallel_save=True,
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/optim/megatron.py:93]: how can this be configured?

File: nemo/lightning/pytorch/strategies.py [hunk 4]
```diff
@@ -541,7 +542,24 @@ def load_model_state_dict(self, checkpoint: Mapping[str, Any], strict: bool = Tr
     @override
     def checkpoint_io(self) -> CheckpointIO:
         if self._checkpoint_io is None:
-            self._checkpoint_io = MegatronCheckpointIO()
+            checkpoint_callback = self.trainer.checkpoint_callback
+            async_save = getattr(checkpoint_callback, "async_save", False)
+            if async_save:
+                self._checkpoint_io = AsyncFinalizableCheckpointIO(
+                    MegatronCheckpointIO(
+                        save_ckpt_format='torch_dist',
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/strategies.py:550]: The `save_ckpt_format`,  `async_save` and `assume_constant_structure` should be configurable too - what's the plan for this?

**Review by mikolajblaz — COMMENTED** [review]
Looks good to me in general, left some comments regarding configurability.

Thanks for including the #9505 changes

File: nemo/lightning/pytorch/strategies.py [hunk 5]
```diff
@@ -541,7 +542,24 @@ def load_model_state_dict(self, checkpoint: Mapping[str, Any], strict: bool = Tr
...
+                        async_save=True,
+                        torch_dist_multiproc=None,  ## TODO: make configurable
+                        assume_constant_structure=False,
+                        parallel_save=True,  ## TODO: make configurable
+                        parallel_load=True,  ## TODO: make configurable
+                    )
+                )
+                self.trainer.callbacks.append(AsyncFinalizerCallback())
+            else:
+                self._checkpoint_io = MegatronCheckpointIO()
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/strategies.py:560]: this should have all the parameters as the async path.
So the logic should be:
```
self._checkpoint_io = MegatronCheckpointIO(<all  the configuration>)
if async_save:
    self._checkpoint_io = AsyncFinalizableCheckpointIO(self._checkpoint_io)
    self.trainer.callbacks.append(AsyncFinalizerCallback())
```

maybe it would be worth checking if AsyncFinalizerCallback() is not already in the callbacks

File: nemo/lightning/pytorch/optim/megatron.py (hunk 3 above)
Comment (ashors1) [reply nemo/lightning/pytorch/optim/megatron.py]: I've made everything configurable from the strategy. Hopefully this is sufficient for now. If there are any concerns, we can work on a better design in a subsequent PR.

File: nemo/lightning/_strategy_lib.py [hunk 6]
```diff
@@ -399,7 +399,7 @@ def enable_nvidia_optimizations() -> None:
 
 
 def optimizer_sharded_state_dict(
-    model: SharedStateDictProtocol, optimizer: "Optimizable", is_loading=False
+    model: SharedStateDictProtocol, optimizer: "Optimizable", is_loading=False, sharding_type='dp_zero_gather_scatter'
```
Comment (mikolajblaz) [inline nemo/lightning/_strategy_lib.py:402]: One last comment, can you make `fully_sharded_model_space` a default sharding type? Thanks
```suggestion
    model: SharedStateDictProtocol, optimizer: "Optimizable", is_loading=False, sharding_type='fully_sharded_model_space'
```

File: nemo/lightning/pytorch/optim/megatron.py [hunk 7]
```diff
@@ -90,10 +90,8 @@ def sharded_state_dict(
                 model_sharded_state_dict,
                 optimizer_state_dict=None,
                 is_loading=False,
-                # dist_ckpt_parallel_save=False, ## TODO: fix!
+                sharding_type='dp_zero_gather_scatter',
```
Comment (mikolajblaz) [inline nemo/lightning/pytorch/optim/megatron.py:93]: same here
```suggestion
                sharding_type='fully_sharded_model_space',
```

**Review by mikolajblaz — DISMISSED** [review]
Thanks for the changes!

One last comment regarding the default sharding type for the optimizer (it's enough to just apply suggestions via browser), other than that all good

File: nemo/lightning/io/pl.py [hunk 8]
```diff
@@ -91,10 +119,14 @@
...
-        dist_checkpointing.save(
-            checkpoint,
-            checkpoint_dir=str(checkpoint_dir),
+        validate_sharding_integrity = not (self.validated_consistency and self.assume_constant_structure)
+        self.validated_consistency = True
+        return dist_checkpointing.save(
+            sharded_state_dict=checkpoint,
+            checkpoint_dir=checkpoint_dir,
             sharded_strategy=self.save_sharded_strategy,
+            validate_access_integrity=validate_sharding_integrity,
```
[the remark this answers is not in the document]
Comment (mikolajblaz) [reply nemo/lightning/io/pl.py:129]: validate_sharding_integrity is always defined 2 lines above

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
